stubs/pluggy/pluggy/_hooks.pyi

```diff
@@ -0,0 +1,117 @@
+from collections.abc import Callable, Generator, Mapping, Sequence, Set as AbstractSet
+from types import ModuleType
+from typing import Any, TypeVar, overload
+from typing_extensions import Final, TypeAlias, TypedDict
+
+from ._result import _Result
+
+_T = TypeVar("_T")
+_F = TypeVar("_F", bound=Callable[..., object])
+_Namespace: TypeAlias = ModuleType | type
+_Plugin = object
+_HookExec: TypeAlias = Callable[[str, Sequence[HookImpl], Mapping[str, object], bool], object | list[object]]
+_HookImplFunction: TypeAlias = Callable[..., _T | Generator[None, _Result[_T], None]]
+
+class _HookSpecOpts(TypedDict):
+    firstresult: bool
+    historic: bool
+    warn_on_impl: Warning | None
+
+class _HookImplOpts(TypedDict):
+    wrapper: bool
+    hookwrapper: bool
+    optionalhook: bool
+    tryfirst: bool
+    trylast: bool
+    specname: str | None
+
+class HookspecMarker:
+    project_name: Final[str]
```

Comment:
`Final` actually has slightly different semantics in a stub file than it does at runtime. While it's okay at runtime to have a "final instance variable", which is what `pluggy` does with this variable:

```py
class HookspecMarker:
    def __init__(self, project_name: str) -> None:
        self.project_name: Final = project_name
```

...in a stub, all variable annotations are put in the class scope, and PEP-591 states:

> Type checkers should infer a final attribute that is initialized in a class body as being a class variable. Variables should not be annotated with both ClassVar and Final.

As such, it doesn't seem to be possible to express the idea of a "final instance variable" in a stub file. Maybe we should just do:

```suggestion
    project_name: str
```

Or maybe we should add a comment:

```suggestion
    # actually a "Final instance variable" at runtime,
    # but that's not expressible in a stub,
    # where all Final variables are implicit ClassVars
    project_name: Final[str]
```

Reply:
> @AlexWaygood thank you for pointing me at PEP 591! Do you think it would be a good idea to add another rule to `flake8-pyi` to guard for it?

Not sure. I'd want to think about it a bit more. In general I feel like the concept of "Final instance variables in stub files" is a bit underspecced, so I'm not sure whether we'd want to enforce a rule or what that rule should enforce 😄